### PR TITLE
reload service account token within incluster config

### PR DESCRIPTION
**Pull request by zshihang**
be default, this will set an expiration period of 1 minutes on service account token. this PR is solving kubernetes-client/ruby/issues/47. as we are graduating the Beta feature Projected Service Account Volume to GA, we expect clients to reload token before it expires.

### CODE REVIEW

**Comment by zshihang** [comment]
@brendandburns  could you also take a look?

File: kubernetes/src/kubernetes/config/incluster_config.rb [hunk 1]
```diff
@@ -62,19 +65,42 @@ def token_file
...
 
+    def token_refresh_period
+      @token_refresh_period ||= TOKEN_REFRESH_PERIOD
+      @token_refresh_period
+    end
+
     def load_token
       File.open(token_file) do |io|
         self.token = io.read.chomp
+        self.token_expires_at = Time.now + token_refresh_period
```
Comment (cben) [inline kubernetes/src/kubernetes/config/incluster_config.rb:76]: Naming: iiuc, the token has an actual validity/expiration time, that's probably not 1 minute, right?  If so, this name may be misleading.
Maybe something like `token_reload_at` or `token_cached_until`...

Comment (cben) [reply kubernetes/src/kubernetes/config/incluster_config.rb:76]: Or maybe use a hidden instance var to not worry about public name :wink:

File: kubernetes/spec/config/incluster_config_spec.rb [hunk 2]
```diff
@@ -50,6 +54,18 @@
 
       incluster_config.configure(actual)
       expect(actual).to be_same_configuration_as(expected)
+
+      old_expired_at = incluster_config.token_expires_at
+      incluster_config.instance_variable_set(
+        :@token_file,
+        Kubernetes::Testing.file_fixture('tokens/token2').to_s
+      )
+      sleep(5)
```
Comment (brendandburns) [inline kubernetes/spec/config/incluster_config_spec.rb:63]: I'd really prefer not to have sleeps in tests. can we inject the clock instead?

File: kubernetes/src/kubernetes/config/incluster_config.rb [hunk 3]
```diff
@@ -62,19 +65,42 @@ def token_file
...
       configuration.host = "#{host}:#{port}"
       configuration.ssl_ca_cert = ca_cert
+      return unless refresh_token
+
+      Configuration.instance_variable_set(:@in_cluster_config, self)
+      Configuration.prepend(Module.new do
+        # rubocop:disable Metrics/LineLength
+        def api_key_with_prefix(identifier)
+          in_cluster_config = self.class.instance_variable_get(:@in_cluster_config)
+          if identifier == 'authorization' && @api_key.key?(identifier) && in_cluster_config.token_expires_at <= Time.now
```
Comment (brendandburns) [inline kubernetes/src/kubernetes/config/incluster_config.rb:95]: The fact that the parameter is `refresh_token` but it's actually conditional is a little weird.

Perhaps name the parameter `try_refresh_token`?

**Comment by brendandburns** [comment]
Minor comments generally looks good to me.

File: kubernetes/src/kubernetes/config/incluster_config.rb (hunk 1 above)
Comment (zshihang) [reply kubernetes/src/kubernetes/config/incluster_config.rb:76]: for the clients, they don't need to see the actual token expiration. it's kinda intentional to have a shorter expiration time exposed to client.

File: kubernetes/spec/config/incluster_config_spec.rb (hunk 2 above)
Comment (zshihang) [reply kubernetes/spec/config/incluster_config_spec.rb:63]: done

File: kubernetes/src/kubernetes/config/incluster_config.rb (hunk 3 above)
Comment (zshihang) [reply kubernetes/src/kubernetes/config/incluster_config.rb:95]: done

**Pull request merged**